contract/src/lib.rs

```diff
@@ -384,23 +384,44 @@ impl MpcContract {
                     PromiseOrValue::Value(signature)
                 }
                 None => {
-                    if depth > 30 {
-                        self.remove_request(&payload);
-                        env::panic_str("Signature was not provided in time. Please, try again.");
+                    // Make sure we have enough gas left to do 1 more call and clean up afterwards
+                    // Observationally 30 calls < 300 TGas so 2 calls < 20 TGas
+                    // We keep one call back so we can cleanup then call panic on the next call
+                    if depth > 29 {
+                        self.fail(
+                            "Signature was not provided in time. Please, try again.",
+                            payload,
+                        )
+                    } else {
+                        log!(&format!(
+                            "sign_helper: signature not ready yet (depth={})",
+                            depth
+                        ));
+                        let account_id = env::current_account_id();
+                        PromiseOrValue::Promise(
+                            Self::ext(account_id).sign_helper(payload, depth + 1),
+                        )
                     }
-                    log!(&format!(
-                        "sign_helper: signature not ready yet (depth={})",
-                        depth
-                    ));
-                    let account_id = env::current_account_id();
-                    PromiseOrValue::Promise(Self::ext(account_id).sign_helper(payload, depth + 1))
                 }
             }
         } else {
-            env::panic_str("unexpected request");
+            self.fail("unexpected request", payload)
```

Comment:
this could also be left as before, since the payload will not exist, and removing it will most probably cause an error

Reply:
Right you are and now fail is only used once so we can inline it.